Clarify onions part 2: a bit deeper rework

rustyrussell opened the pull request:
This is based on #1181 (first four commits), and replaces #1179 based on feedback there.

You're going to want to review one commit at a time: I rewrote it to be managable to review.

1. Move Route Blinding Section up a level, put blinded_path type there
    - It's not just for onions any more
2. Update path construction requirements to explicitly refer to `blinded_path` elements
    - It's now an actual type, so we can name fields.
3. Requirements for *using* a blinded path (esp. payments where we need to route non-blinded to the introduction node) as well as forwarding them.
4. Attempt to merge @t-bast and my high-level description
5. Add dummy hops as a MAY in requirements

Basically, nothing changes, but hopefully it will be more digestable in future!

t-bast commented on 04-onion-routing.md:
```diff
@@ -432,124 +433,114 @@ otherwise meets the amount criterion (eg. some other failure, or
+1. an initial introduction point (`first_node_id`)
+2. an initial key to share a secret with the first node_id (`first_path_key`)
+3. a series of tweaked node ids (`path.blinded_node_id`)
+4. a series binary blobs encrypted to the nodes (`path.encrypted_recipient_data`)
```
nit:
```suggestion
4. a series of binary blobs encrypted to the nodes (`path.encrypted_recipient_data`)
```

t-bast commented on 04-onion-routing.md:
```diff
@@ -432,124 +433,114 @@ otherwise meets the amount criterion (eg. some other failure, or
-- MUST create a blinded node ID $`B_i`$ for each node using the following algorithm:
+- MUST create a viable path to itself ($`N_r`$) i.e. $`N_0 \rightarrow N_1 \rightarrow ... \rightarrow N_r`$.
+- MUST set `first_node_id` to $`N_0`$
+- MUST create a series of ECDH shares secrets for each node in the route using the following algorithm:
```
```suggestion
- MUST create a series of ECDH shared secrets for each node in the route using the following algorithm:
```

t-bast commented on 04-onion-routing.md:
```diff
@@ -432,124 +433,114 @@ otherwise meets the amount criterion (eg. some other failure, or
+- MUST set `first_path_key` to $`E_0`$
+- MUST create a series of blinded node IDs $`B_i`$ for each node using the following algorithm:
+  - $`B_i = HMAC256(\text{"blinded\_node\_id"}, ss_i) * N_i`$ (blinded `node_id` for $`N_i`$, private key known only by $`N_i`$)
+  - MUST set `blinded_node_id` for each `onionmsg_hop` in `path` to $`B_i`$
```
If we rename `onionmsg_hop`, this should be renamed here as well.

t-bast commented on 04-onion-routing.md:
```diff
@@ -432,124 +433,114 @@ otherwise meets the amount criterion (eg. some other failure, or
-  decrypted fields to locate the next node
-- If the `encrypted_data` field is missing or cannot be decrypted:
+- MUST decrypt the `encrypted_recipient_data` field using $`rho_i`$ as a key using ChaCha20-Poly1305 and an all-zero nonce key.
+- If the `encrypted_recipient_data` field is missing cannot be decrypted into an `encrypted_data_tlv` or contains unknown even fields:
```
```suggestion
- If the `encrypted_recipient_data` field is missing, cannot be decrypted into an `encrypted_data_tlv` or contains unknown even fields:
```

t-bast reviewed (commented):
Thanks, this is now much better! The individual commits are easy to follow and all make sense, and the end result looks good to me!

rustyrussell commented on 04-onion-routing.md:
```diff
@@ -432,7 +433,20 @@ otherwise meets the amount criterion (eg. some other failure, or
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_hop`:`path`]
+
+1. subtype: `onionmsg_hop`
```
Yes, I like this.  To avoid confusion, I'll apply this as a new commit at the end.

rustyrussell commented:
OK:
1. rebased on master (now #1181 is applied)
2. Applied @t-bast typo fixes
3. Renamed onionmsg_hop to blinded_path_hop as per @t-bast recommendation

t-bast reviewed (commented):
Thanks, this looks good! Can you also update `route-blinding-test.json`? It contains fields called `ephemeral_pubkey` and `ephemeral_privkey`, which is confusing and should be replaced by `path_key` and `path_privkey`.

gijswijs commented on 04-onion-routing.md:
```diff
@@ -599,6 +589,48 @@ set `payment_constraints.max_cltv_expiry` to restrict the lifetime of a blinded
 
+### Inside `encrypted_recipient_data`: `encrypted_data_tlv`
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
```
Use `encrypted_recipient_data` here as well, for consistency.

rustyrussell replied:
Replaced everywhere, thanks!

gijswijs commented on 04-onion-routing.md:
```diff
@@ -599,6 +589,48 @@ set `payment_constraints.max_cltv_expiry` to restrict the lifetime of a blinded
+
+#### Rationale
+
+Encrypted recipient data is created by the final recipient to give to the
```
IIUC in the case of 'reply_path` for onion messaging, it's not the recipient that gives the payer this data, but the sender giving the receiver this data. Because route blinding is now a more generic tool used for more than payments, we should probably refrain from the usage of `payer` and lingo alike.

rustyrussell replied:
Good catch!  I reworked this:

Encrypted recipient data is created by the final recipient to give to the
sender, containing instructions for the node on how to handle the message (it can also be created by the sender themselves: the node forwarding cannot tell).  It's used
in both payment onions and onion messages onions.  See [Route Blinding](#route-blinding).

t-bast commented on 04-onion-routing.md:
```diff
@@ -1475,19 +1507,6 @@ even, of course!).
-
 #### Requirements
 
 The creator of `encrypted_recipient_data` (usually, the recipient of the onion):
```
> In the case of Onion Messages the creator of encrypted_recipient_data isn't usually the recipient of the onion.

With onion messages, you're still using blinded paths, so the recipient of the onion does create some `encrypted_recipient_data`. Then the sender of the onion message will also create encrypted data for the hops it adds *before* the blinded path's introduction point.

rustyrussell replied:
Yeah, we could argue over "usually", but it's a good mental model.  It's designed to be *able* to be generated by the recipient, and the sender can fake it...

rustyrussell commented on 04-onion-routing.md:
```diff
@@ -476,61 +476,62 @@ initial blinding point, and have it sent to the introduction node.
+- For the first entry in `path`:
+  - if it is sending a payment:
+    - MAY encrypt the first blinded path onion to `first_node_id` and include `first_path_key` as `current_path_key`.
+    - MUST use this method if the prior node does not support `option_route_blinding`.
```
I was thinking that we might not want the introductory node to know it's being used that way.  However, for now it's fine to require this and put a note in the rationale that using a prior hop is possible.

t-bast reviewed (approved):
ACK https://github.com/lightning/bolts/pull/1182/commits/0aae209c1a69437fca3e20bd013dc0cfac5f77f2

t-bast commented:
@Roasbeef @arik-so @jkczyz could you please take a look at this? It would be nice to land it before the next spec meeting.

valentinewallace reviewed (approved):
Reads a lot more clearly to me! The new section on path concatenation clarifies a lot. A good amount of the additions to BOLT 4 seem a bit redundant to what's in `proposals/route_blinding.me` but that seems ok.

t-bast commented:
@Roasbeef can you take a quick look at this PR? As discussed during yesterday's spec meeting, we should merge this once we have your ACK :)

Roasbeef commented on 04-onion-routing.md:
```diff
@@ -476,61 +476,62 @@ initial blinding point, and have it sent to the introduction node.
 
+The writer of a `blinded_path`:
+
+- MUST create a viable path to itself ($`N_r`$) i.e. $`N_0 \rightarrow N_1 \rightarrow ... \rightarrow N_r`$.
```
😎

Roasbeef reviewed (approved):
LGTM 🎫
